Assignment3/Breakout.java

```diff
@@ -0,0 +1,252 @@
+			} else if (i <= 10) {
+				currentColor = Color.CYAN;
+			}
+			for (int j = 1; j <= NBRICKS_PER_ROW; j++) {
+
+				buildBrick(currentColor);
+				currentBrickLeftSide = currentBrickLeftSide + BRICK_WIDTH + BRICK_SEP;
+			}
+			currentBrickLeftSide = BRICK_SEP;
+		}
+	}
+	
+	/**
+     * Defines a unified format of a brick. Ensures that all bricks are the same size and can be colored.
+     */
+
+	public void buildBrick(Color c) {
+		brick = new GRect(currentBrickLeftSide, bottomLineOfCurrentRow, BRICK_WIDTH, BRICK_HEIGHT);
+		add(brick);
+		brick.setFilled(true);
+		brick.setColor(c);
+	}
+
+	public void createPaddle() {
+		paddle = new GRect(WIDTH / 2 - PADDLE_WIDTH / 2, HEIGHT - PADDLE_Y_OFFSET, PADDLE_WIDTH, PADDLE_HEIGHT);
+		add(paddle);
+		paddle.setFilled(true);
+	}
+
+	public void createBall() {
+		ball = new GOval(WIDTH / 2, HEIGHT / 2, BALL_RADIUS, BALL_RADIUS);
+		add(ball);
+		ball.setColor(Color.GREEN);
+		ball.setFilled(true);
+	}
+	
+	/**
+     * Defines the way that ball will be moving. 
+     */
+	
+	private void moveBall() {
+		vx = rgen.nextDouble(1.0, 3.0);
+		if (rgen.nextBoolean(0.5))
+			vx = -vx;//makes the ball bounce off the object in the opposite vertical direction
+		ball.move(vx, vy);
+		if (ballHitTheVerticallWall()) {
+			vx = -vx;//makes the ball bounce off the object in the opposite vertical direction
+		}
+		if (ballHitTheTopWall()) {
+			vy = -vy;//makes the ball bounce off the object in the opposite vertical direction
+		}
+	}
+	
+	/**
+     * Enables paddle to react to mouse movement only horizontally and only within the game board.
+     */
+
+	public void mouseMoved(MouseEvent e) {
+		if (e.getX() <= WIDTH && e.getX() >= 0 + PADDLE_WIDTH) {
+			paddle.setLocation(e.getX() - PADDLE_WIDTH, HEIGHT - PADDLE_Y_OFFSET);
+		}
+	}
+
+	/**
+     * Let's user to make the ball appear on click only. Also with a click user can hide any text in the middle.
+     */
+	
+	public void mouseClicked(MouseEvent e) {
+		if (ball == null && remainingLifeCount >= 0) {
+			createBall();
+		}
+		if (remainingLifeCount == NTURNS && middleText.isVisible()) {
+			remove(middleText);
+		}
+		if (remainingLifeCount < NTURNS && lostLife.isVisible()) {
+			remove(lostLife);
+		}
+	}
+
+	public void checkForCollisions() {
+		getCollidingObjects();
+		if (isThereACollider() && (collider != paddle)) {
+			remove(collider);
+			vy = -vy;
+			remainingBricks = remainingBricks - 1;
+		} else if (collider == paddle) {
+			vy = -vy;
+		}
+	}
+
+	private boolean ballHitTheVerticallWall() {
+		boolean hitVertical = ball.getX() <= (WIDTH - BALL_RADIUS) || ball.getX() >= 0 + BALL_RADIUS;
+		return hitVertical;
+	}
+
+	private boolean ballHitTheTopWall() {
+		boolean hitTop = ball.getY() <= 0;
+		return hitTop;
+	}
+
+	private boolean isThereACollider() {
+		if (collider != null) {
+			return true;
+		} else
+			return false;
+	}
+
+	private GObject getCollidingObjects() {
+		return collider = getElementAt(ball.getX(), ball.getY());
```

Comment:
This only checks top corner of collision. Let's try to be more comprehensive.

Reply:
If we have check for collision in moveBall we won't need a global collision object.